pynxtools_xps/phi/spe_pro_phi.py

```diff
@@ -0,0 +1,1549 @@
+    "xray_offset",
+    "xray_rotation",
+    "xray_emission_current",
+    "xray_max_filament_current",
+    "xray_stigmator_x",
+    "xray_stigmator_y",
+    "flood_gun_current",
+    "flood_gun_energy",
+    "flood_gun_extractor",
+    "flood_gun_filament_current",
+    "flood_gun_pulse_length",
+    "flood_gun_pulse_frequency",
+    "flood_gun_time_per_step",
+    "flood_gun_ramp_rate",
+    "flood_gun_x_steering",
+    "flood_gun_y_steering",
+    "sxi_binding_energy",
+    "sxi_pass_energy",
+    "sxi_lens2_voltage",
+    "sxi_lens3_voltage",
+    "sxi_lens4_voltage",
+    "sxi_lens5_voltage",
+    "sxi_rotator",
+    "sxi_lens_bias_voltage",
+    "sxi_shutter_bias_voltage",
+    "detector_acquisition_time",
+    "stage_current_rotation_speed",
+    "defect_positioner_u",
+    "defect_positioner_v",
+    "defect_positioner_x",
+    "defect_positioner_y",
+    "defect_positioner_z",
+    "defect_positioner_tilt",
+    "defect_positioner_rotation",
+    "gcib_sputter_rate",
+    "gcib_high_voltage",
+    "gcib_ionization",
+    "gcib_extractor",
+    "gcib_wien_filter_voltage",
+    "gcib_bend_voltage",
+    "gcib_emission",
+    "gcib_magnet_current",
+    "gcib_focus",
+    "gcib_objective",
+    "gcib_focus",
+    "gcib_gas_pressure",
+    "gcib_cluster_size",
+    "gcib_energy_per_atom",
+    "deconvolution_pass_energy",
+]
+
+UNIT_MISSING: Dict[str, str] = {
+    "grid_voltage": "V",
+    "condenser_lens_voltage": "V",
+    "objective_lens_voltage": "V",
+    "bend_voltage": "V",
+    "neutral_grid_voltage": "V",
+    "neutral_condenser_lens_voltage": "V",
+    "neutral_objective_lens_voltage": "V",
+    "stage_current_rotation_speed": "degree/s",
+    "neutral_bend_voltage": "V",
+    "defect_positioner_u": "mm",
+    "defect_positioner_v": "mm",
+    "defect_positioner_x": "mm",
+    "defect_positioner_y": "mm",
+    "defect_positioner_z": "mm",
+    "defect_positioner_tilt": "degree",
+    "defect_positioner_rotation": "degree",
+    "flood_gun_pulse_frequency": "1/s",
+    "flood_gun_x_steering": "mm",
+    "flood_gun_y_steering": "mm",
+    "profiling_sputter_delay": "s",
+    "scan_deflection_span_x": "mm",
+    "scan_deflection_span_y": "mm",
+    "scan_deflection_offset_x": "mm",
+    "scan_deflection_offset_y": "mm",
+    "xray_stigmator_x": "mm",
+    "xray_stigmator_y": "mm",
+}
+
+
+class MapperPhi(XPSMapper):
+    """
+    Class for restructuring .xy data file from
+    Phi vendor into python dictionary.
+    """
+
+    config_file = "config_phi.json"
+
+    def __init__(self):
+        super().__init__()
+        self.write_channels_to_data = True
+
+    def _select_parser(self):
+        """Select the proper Phi data parser."""
+        return PhiParser()
+
+    def construct_data(self):
+        """Map Phi format to NXmpes-ready dict."""
+        # pylint: disable=duplicate-code
+        spectra = copy.deepcopy(self.raw_data)
+
+        self._xps_dict["data"]: dict = {}
+
+        key_map: Dict[str, List[str]] = {
```

Comment:
Should this not be coming from the config file?

Reply:
This comes before the config file is even parsed. I generally use a `mapper` layer that is called by the general XPS reader and acts as an interface to different sub-parsers (i.e., for different software versions). This mapper then creates from the parser data a dictionary. This dict is used by the main parser to fill the template through the config file (with a token, similarly to what you have in the multi-format reader). The advantage of having this mapper layer is that I get to keep some functionality across the different file formats, so that the config file and filling of the template are always relatively similar. See also my comment [here](https://github.com/FAIRmat-NFDI/pynxtools/pull/250#issuecomment-1959742939).

It is a bit inefficient because I first parse the data (on the parser level) in a list of dicts (one for each spectrum) and then I populate the large mapper dict from there (though the performance is not that bad). In general, the parsers should create such dicts immediately. Eventually, I want to change it (see issue #19), but I will only do it once I switch to the multi-format reader.

Reply:
Ok, makes sense. I was just a bit confused when I read it but I think it's totally fine. So I'll just resolve this